src/entities/join/weighted/weightedJoin.ts

```diff
@@ -0,0 +1,149 @@
+import { createPublicClient, encodeFunctionData, http } from 'viem';
+import { Token, TokenAmount, WeightedEncoder } from '../../..';
+import { Address } from '../../../types';
+import {
+    BALANCER_HELPERS,
+    BALANCER_VAULT,
+    CHAINS,
+    ZERO_ADDRESS,
+} from '../../../utils';
+import { balancerHelpersAbi, vaultAbi } from '../../../abi';
+import {
+    checkInputs,
+    getAmountsIn,
+    getAssets,
+    getJoinParameters,
+    getUserData,
+} from './helpers';
+import {
+    BaseJoin,
+    JoinCallInput,
+    JoinInput,
+    JoinKind,
+    JoinQueryResult,
+    PoolState,
+} from '..';
+
+export class WeightedJoin implements BaseJoin {
+    // TODO - Probably not needed
+    getInstance(): WeightedJoin {
+        return new WeightedJoin();
+    }
+
+    public async query(
+        input: JoinInput,
```

Comment:
would generally maybe find a more explicit way to deal with native asset joins. Perhaps a flag on `BaseJoinInput` called something like `joinWithNativeAsset?: boolean` and not expecting a zero address, as it is another hidden requirement that is easy to overlook.

Reply:
yeah, I'm so used to think of native asset joins with zero address that I didn't think of that - I agree it's a hidden requirement that we can make more explicit.
I'm assuming that with that approach we would expect the `wrapped` asset as input and replace that with `zero address` if `joinWithNativeAsset = true`, right?
Would we also accept `zero address` as input? Or should we throw if the user tries that?
I ask because limiting the input to pool assets only and not having to handle the case where `zero address` is passed would reduce a lot the complexity there.